triage/summarize/map_abstractions.go

```diff
@@ -0,0 +1,209 @@
+key-value pairs as a map.
+*/
+
+package summarize
+
+import (
+	"sort"
+)
+
+// map[string][]failure type alias
+
+// failuresGroup maps strings to failure slices.
+type failuresGroup map[string][]failure
+
+// failuresGroupPair is a representation of a failuresGroup key-value mapping as a two-element
+// struct.
+type failuresGroupPair struct {
+	Key      string    `json:"key"`
+	Failures []failure `json:"failures"`
+}
+
+// keys provides the failuresGroup's keys as a string slice.
+func (fg *failuresGroup) keys() []string {
+	result := make([]string, len(*fg))
+
+	iter := 0
+	for key := range *fg {
+		result[iter] = key
+		iter++
+	}
+
+	return result
+}
+
+// asSlice returns the failuresGroup as a failuresGroupPair slice.
+func (fg *failuresGroup) asSlice() []failuresGroupPair {
+	result := make([]failuresGroupPair, len(*fg))
+
+	iter := 0
+	for str, failures := range *fg {
+		result[iter] = failuresGroupPair{str, failures}
+		iter++
+	}
+
+	return result
+}
+
+// sortByNumberOfFailures returns a failuresGroupPair slice sorted by the number of failures in each
+// pair, descending. If the number of failures is the same for two pairs, they are sorted alphabetically
+// by their keys.
+func (fg *failuresGroup) sortByNumberOfFailures() []failuresGroupPair {
+	result := fg.asSlice()
+
+	// Sort the slice.
+	sort.Slice(result, func(i, j int) bool {
+		iFailures := len(result[i].Failures)
+		jFailures := len(result[j].Failures)
+
+		if iFailures == jFailures {
+			return result[i].Key < result[j].Key
+		}
+
+		// Use > instead of < so the largest values (i.e. clusters with the most failures) are first.
+		return iFailures > jFailures
+	})
+
+	return result
+}
+
+// equal determines whether this failuresGroup is deeply equal to another failuresGroup.
+func (a *failuresGroup) equal(b *failuresGroup) bool {
+	// First check the length to deal with different-length maps
+	if len(*a) != len(*b) {
+		return false
+	}
+
+	for key, failuresA := range *a {
+		// Make sure the other map contains the same keys
+		if failuresB, ok := (*b)[key]; ok {
+			// Check lengths
+			if len(failuresA) != len(failuresB) {
+				return false
+			}
+			// Compare the failures slices
+			for i := range failuresA {
+				if failuresA[i] != failuresB[i] {
+					return false
+				}
+			}
+		} else {
+			// The other map is missing a key
+			return false
+		}
+	}
+
+	return true
+}
+
+// map[string]failuresGroup type alias, which is really a map[string]map[string][]failure type alias
+
+// nestedFailuresGroups maps strings to failuresGroup instances.
+type nestedFailuresGroups map[string]failuresGroup
+
+// nestedFailuresGroupsPair is a representation of a nestedFailuresGroups key-value mapping as a
+// two-element struct.
+type nestedFailuresGroupsPair struct {
+	Key   string        `json:"key"`
+	Group failuresGroup `json:"group"`
+}
+
+// keys provides the nestedFailuresGroups's keys as a string slice.
+func (nfg *nestedFailuresGroups) keys() []string {
```

Comment:
I have seen a few of these can it be made generic? Unsure if this is an option
```
(fg *map[string]interface{}) keys()
```

Reply:
From what I've seen, no. Even using a type switch results in repeated code